contracts/signature/SignatureChecker.sol

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.0 <0.9.0;
+
+import "./IERC1271.sol";
+
+/// @title SignatureChecker - A contract that extracts and inspects signatures appended to calldata.
+/// @notice currently supporting eip-712 and eip-1271 signatures
+abstract contract SignatureChecker {
+    uint256 private nonce;
+
+    /**
+     * @dev Returns the current nonce value.
+     */
+    function moduleTxNonce() public view returns (uint256) {
+        return nonce;
+    }
+
+    /**
+     * @dev Increments nonce.
+     */
+    function moduleTxNonceBump() internal {
+        nonce = nonce + 1;
+    }
+
+    /**
+     * @dev When signature present in calldata, returns the address of the signer.
+     */
+    function moduleTxSignedBy() internal view returns (address signer) {
+        bytes calldata data = msg.data;
+        if (data.length >= 4 + 65) {
+            (
+                bytes32 r,
+                bytes32 s,
+                uint8 v,
+                bool isEIP1271,
+                uint256 start
+            ) = _splitSignature(data);
+
+            bytes32 hash = _moduleTxHash(data[:start], nonce);
+
+            if (isEIP1271) {
+                // When handling contract signatures the address
+                // of the contract is encoded into r
+                signer = address(uint160(uint256(r)));
+                uint256 end = data.length - 65;
+
+                if (!_isValidContractSignature(signer, hash, data[start:end])) {
+                    signer = address(0);
+                }
+            } else {
+                signer = ecrecover(hash, v, r, s);
+            }
+        }
+    }
+
+    /**
+     * @dev Extracts signature from calldata, and divides it into `uint8 v, bytes32 r, bytes32 s`.
+     * @param data current transaction calldata.
+     */
+    function _splitSignature(
+        bytes calldata data
+    )
+        private
+        pure
+        returns (bytes32 r, bytes32 s, uint8 v, bool isEIP1271, uint256 start)
+    {
+        uint256 length = data.length;
+        r = bytes32(data[length - 65:]);
+        s = bytes32(data[length - 33:]);
+        v = uint8(bytes1(data[length - 1:]));
+        isEIP1271 = v == 0 && (uint256(s) < (length - 65));
+        start = isEIP1271 ? uint256(s) : length - 65;
+    }
+
+    /**
+     * @dev Hashes the EIP-712 data structure that will be signed by owner.
+     * @param data The data for the transaction.
+     * @param _nonce The nonce value.
+     * @return the bytes32 hash to be signed by owners.
+     */
+    function _moduleTxHash(
+        bytes calldata data,
+        uint256 _nonce
+    ) private view returns (bytes32) {
```

Comment:
Could make this public so that it can be easily queried? This way one doesn't need to replicate the hash function elsewhere.

Reply:
Sorted